packages/@aws-cdk/cdk/lib/physical-name.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * Represents the human-readable, physical name of a resource.
+ * These can be known at code runtime,
+ * in the case of fixed, customer-provided names,
+ * at synthesis time, in case of automatically assigned names by the framework,
+ * or only at deploy time, assigned by CloudFormation -
+ * which is the default.
+ *
+ * @see #deployTime()
+ * @see #deployTimeOrAssigned()
+ * @see #fixed()
+ */
+export abstract class PhysicalName {
+  /**
+   * A physical name that will be automatically assigned at deploy time.
+   * This is the default you don't specify a physical name.
+   */
+  public static deployTime(): PhysicalName {
+    return new DeployTimePhysicalName();
+  }
+
+  /**
+   * A physical name that will be assigned at synthesis time if the resource
+   * is used in a cross-environment manner,
+   * and at deploy time otherwise.
+   * This is the default name implicitly created sub-resources
+   * (like IAM Roles) are created with in the Construct Library.
+   * If you're using a Construct in a cross-environment manner,
+   * you need to use either this name,
+   * or a {@link #fixed() fixed name}.
+   */
+  public static deployTimeOrAssigned(): PhysicalName {
+    return new DeployTimeOrAssignedPhysicalName();
+  }
+
+  /**
+   * A fixed physical name (one that is known statically, at synthesis time).
+   *
+   * @param name the name to assign
+   */
+  public static fixed(name: string): PhysicalName {
+    return new FixedPhysicalName(name);
+  }
+
+  public abstract availableAtSynthesisTime: boolean;
+
+  protected constructor() {
+  }
+
+  public abstract asString(): string | undefined;
+}
+
+// implementations are private to this module,
+// we only surface them through static factory methods
+
+class DeployTimePhysicalName extends PhysicalName {
+  public readonly availableAtSynthesisTime = false;
+
+  public asString(): string | undefined {
+    return undefined;
```

Comment:
Shouldn't that be an exception?

Reply:
Not really, no - if a customer did not provide an explicit name, we should return `undefined` (this way this can be used for L1 props: `new CfnBucket(this, 'Resource', { bucketName: this.physicalName.asString() });`).

Reply:
So maybe `asString` is not a good name for this method. Maybe just `.value()`? A physical name is always a string. Also, why isn't this a property? (`physicalName.value` sounds good).

Reply:
Will change to `value`, and a property.

Reply:
Changed.